rescript/get_data.py

```diff
@@ -12,13 +12,132 @@
 import shutil
 import gzip
 import warnings
+import requests
+import tarfile
 
 import qiime2
 from urllib.request import urlretrieve
 from urllib.error import HTTPError
 from q2_types.feature_data import RNAFASTAFormat
 
 
+def _unite_dois_to_urls(DOIs):
+    '''Generate UNITE urls, given their DOIs.'''
+    # Make DOIs iterable
+    DOIs = [DOIs] if isinstance(DOIs, str) else DOIs
+    print('Get URLs for these DOIs:', DOIs)
+    base_url = 'https://api.plutof.ut.ee/v1/public/dois/'\
+               '?format=vnd.api%2Bjson&identifier='
+    # Eventual output
+    URLs = set()
+    # For each DOI, get download URL of file
+    for DOI in DOIs:
+        query_data = requests.get(base_url + DOI).json()
+        # Updates can be made to files in a DOI, so on the advice of the devs,
+        # only return the last file uploaded with this -1  vv
+        URL = query_data['data'][0]['attributes']['media'][-1]['url']
+        URLs.add(URL)
+    return URLs
+
+
+def _unite_get_url(version, taxon_group, singletons):
+    '''Generate UNITE urls, given database version and reference target.'''
+    # Lookup DOIs for various databases, source: https://unite.ut.ee/repository.php
+    unite_dois = {
+        '9.0': {'fungi':      {False: '10.15156/BIO/2938079', True: '10.15156/BIO/2938080'},
+                'eukaryotes': {False: '10.15156/BIO/2938081', True: '10.15156/BIO/2938082'}},
+        # Old version 9.0 is not listed here
+        '8.3': {'fungi':      {False: '10.15156/BIO/1264708', True: '10.15156/BIO/1264763'},
+                'eukaryotes': {False: '10.15156/BIO/1264819', True: '10.15156/BIO/1264861'}},
+        '8.2': {'fungi':      {False: '10.15156/BIO/786385', True: '10.15156/BIO/786387'},
+                'eukaryotes': {False: '10.15156/BIO/786386', True: '10.15156/BIO/786388'}},
+        '8.0': {'fungi':      {False: '', True: '10.15156/BIO/786349'},
+                'eukaryotes': {False: '', True: ''}},  # All other 8.0 are in zip files
+    }
+    # There's got to be a better way! See https://stackoverflow.com/questions/25833613/safe-method-to-get-value-of-nested-dictionary
+    try:
+        # Check if we have the DOI requested
+        target_doi = unite_dois[version][taxon_group][singletons]
+    except KeyError as ke:
+        print('Unknown DOI for this value: ' + str(ke))
```

Comment:
raise a warning instead of using a print statement.

We should never reach this point anyway unless if UNITE shuffles something around or if the URL request somehow failed. So I would suggesting printing something like "URL request failed for ...."

Reply:
oh, I intended this `except` to catch typos
`_unite_get_url(version = '9', taxon_group = 'Fungi')`

What's the best way to catch typos for things we don't have in our DOI dictionary?

Reply:
QIIME 2 would catch the typos (because you would hardcode the acceptable dates/versions etc in the plugin registration). So you do not need to worry about typos internally.